More careful calculation of inclusive range limits (fixes #105)

mattsep opened the pull request:
In the `range_count` function, floating point error would lead to inaccurate calculation of the size of an inclusive range. 

In the case of the example shown in #105, with `start = 0.1`, `end = 2.0`, and `step = 0.1`, we compute the number of elements in the range to be `1 + (int) ((2.0 - 0.1) / 0.1) = 1 + (int) (18.99...96) = 19` which is off by one. Instead, we need to make sure to take the ceiling of the result before casting to an integer.

I additionally removed a possible integer division by 0, as well as added checks for infinities and nans before casting to an integer. In these cases, I take the number of elements to be 0 since there's no well-defined set of elements to return.

This should resolve #105

mattsep commented:
The use of `DBL_EPSILON` is probably a clue that there are limits to the validity of this fix. Inclusive ranges with floats are a bit fragile since they rely on a test for _equality_ on the right bound, which is a bit fragile.

There is an additional hidden bug in the parsing of exclusive ranges that this PR does not resolve, as showcased by this example:
```
print List(1.0 .. 1.2 : 0.15) // prints [ 1 ], but should print [ 1, 1.15 ]
```
This happens because the exclusive range `1.0 .. 1.2 : 0.15` is converted to `1.0 .. 1.05 : 0.15` by subtracting the step from the end, which is not the right thing to do in this situation. Since this happens at parse time, it's not possible to revert this calculation in the `range_count` function.

If we were to instead convert all inclusive ranges to exclusive ranges by adding the step to the end, we would find that `1.0 .. 1.2 : 0.15` would be converted to `1.0 ... 1.35 : 0.15` which would print `[ 1, 1.15, 1.3 ]` and this is _also_ not correct.

What we need is to explicitly flag whether a given range is inclusive or exclusive range at construction so that we can compute the correct number of elements in the `range_count` function, at which point we can convert both ranges to one or the other.

softmattertheory commented:
This is a very good catch @mattsep! Thanks for this and we can move the above to an issue.

ConduitDan commented:
Great find @mattsep.
Could you write a test for this?

mattsep commented:
Can do. Should I preemptively add a test case for the failing exclusive range example I show?

ConduitDan commented:
as a note: the test range/constructor.morpho failed in your branch

I have pulled a change that makes the CI script actually report these failures

mattsep commented:
I added some tests, some of which use the `AssumeFail` syntax (which are the ones that fail in CI). I fixed the issue causing the`range/constructor.morpho` test to fail (forgot to take an absolute value, which made negative ranges not work).

ConduitDan commented:
Awesome thanks! I'll implement assumeFail this afternoon

ConduitDan commented:
On second thought lets just remove the test for the new issue now and just make sure we capture it in an issue.

ConduitDan commented:
Does this change introduce the 
`print List(1.0 .. 1.2 : 0.15) // prints [ 1 ], but should print [ 1, 1.15 ]`
bug?

mattsep commented:
Whoops, sorry Dan, that was a typo! Should be `print List(1.0 ... 1.2 : 0.15)` (note the three dots)

ConduitDan commented:
got it! Lets remove the AssumeFail test as its caputred by #114. I think we can then go ahead an merge this in

mattsep commented:
Sorry for the delay! Tests related to #114 have been removed, should be good to merge

softmattertheory commented:
If we're ready to go, @ConduitDan, feel free to merge!
